### PR TITLE
fix: use dynamic import of @octokit/rest to handle differences with E…

### LINKED ISSUE #452: Fix require() synxtax issue with Octokit in github.js
These logs were found indicating an issue:

Screenshot 2024-09-23 at 17 19 01

The issue is related to mixing ESM and CommonJS modules together in one project incorrectly. We should upgrade the project to use ESM as part of the fix.

### PR DESCRIPTION
**Pull request by nickdasilva8**
…SM and Commonjs

fixes #452

~~Based on [this](https://nodejs.org/api/esm.html#esm_import_expressions) documentation, and local testing with importing the module within a function where it was dynamically imported, this PR should resolve the issue.~~

- Now an ESM app
- Converted tests to use node:test instead of Tap
- Require node 20+
- Added test that imports all files used by index.js to catch import/require issues.

### CODE REVIEW

**Review by simoneb — COMMENTED** [review]
it would have probably been nicer to switch the app to esm. have you considered that?

**Comment by nickdasilva8** [comment]
> it would have probably been nicer to switch the app to esm. have you considered that?

@simoneb I did consider it. I wasn't sure on the urgency of this fix, so assumed a fix without additional work would be the best approach. Upgrading to ESM isn't require to fix the current issue. I figured fixing the problem properly (this isn't a hacky approach in my mind) and quickly is a win-win.

If there's no urgency on this fix, theb happy to spend the additional time upgrading the app to ESM - I just didn't know where an upgrade to ESM sat inbetween everything else you have lined up for me to work on.

**Comment by simoneb** [comment]
not urgent, GCP keeps the old revision running if the new one isn't starting. so let's take the time to do the proper fix

**Comment by nickdasilva8** [comment]
Alright, sounds good. I'll update it. :)

File: index.test.js [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
+
+import setup from './test-setup.js';
+
+test('Ensure all imports used by index.js are correct', async t => {
+  try {
+    await import('./lib/github.js')
+    await import('./lib/auth.js')
+    await import('./lib/routes.js')
+    
+    assert.ok(true, 'all imports used by index.js import correctly')
```
Comment (simoneb) [inline index.test.js:13]: not mandatory, but you can use `t.assert` everywhere

**Review by simoneb — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -2,11 +2,13 @@
   "name": "optic-release-automation-app",
   "version": "0.4.26",
   "description": "",
-  "main": "index.js",
+  "exports": "./index.js",
+  "type": "module",
+  "node": ">=20",
```
Comment (nickdasilva8) [inline package.json:7]: @simoneb is restricting this to node 20+ a problem? I went off the work we were doing in Fastify.

File: index.test.js (hunk 1 above)
Comment (nickdasilva8) [reply index.test.js:13]: Happy with it either way, if you'd prefer I update it just thumbs up this comment and I'll push up the change.

File: package.json (hunk 2 above)
Comment (simoneb) [reply package.json:7]: I don't think so, but let's make sure we're deploying to a recent version of Node in cloud run. See the cd workflow, we don't seem to be specifying a version there

File: index.test.js (hunk 1 above)
Comment (simoneb) [reply index.test.js:13]: no need, more of a FYI

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -27,6 +27,10 @@ jobs:
         with:
           workload_identity_provider: ${{ vars.GCP_WORKLOAD_IDENTITY_PROVIDER }}
           service_account: ${{ vars.GCP_SERVICE_ACCOUNT }}
+      - name: Set up Node.js
```
Comment (simoneb) [inline .github/workflows/cd.yml:30]: this is not actually needed, what I meant is that we don't specify a runtime in the deploy-cloudrun action below, so I guess it defaults to either what's the latest LTS they support, probably adhering to the engines field in package.json. maybe there's nothing we need to do here actually, just worth checking what version we're running on atm and making sure that when this is merged and deployed, it runs on a compatible node version

File: .nvmrc [hunk 4]
```diff
@@ -1 +1 @@
-19
+20
```
Comment (simoneb) [inline .nvmrc:1]: let's use `lts/*`

**Pull request merged**